reps/2024-10-18-train-tune-api-revamp/2024-10-18-train-tune-api-revamp.md

```diff
@@ -0,0 +1,1194 @@
+
+<table>
+<!-- Table Headers -->
+  <tr>
+    <th>Change Type</th>
+    <th>Diff</th>
+    <th>Notes</th>
+  </tr>
+  <tr>
+<td>🚧</td>
+<td>
+
+```diff
+-Tuner(
+-   trainable: Union[
+-       ray.tune.Trainable, ray.train.BaseTrainer
+-   ] = TorchTrainer(...)
+-)
++Tuner(trainable: ray.tune.Trainable = train_driver_fn)
+```
+
+</td>
+<td>
+
+The input to the Tuner is now restricted to Trainable types, instead of allowing the `Trainer` type as a special case. See [here](#hyperparameter-optimization) for an example.
+
+</td>
+  </tr>
+
+  <tr>
+<td>🔀</td>
+<td>
+
+```diff
+ray.tune.Tuner(
+-   run_config: ray.train.RunConfig(
+-       checkpoint_config: ray.train.RunConfig,
+-       sync_config: ray.train.RunConfig,
+-       failure_config: ray.train.RunConfig,
+-   )
++   run_config: ray.tune.RunConfig(
++       checkpoint_config: ray.tune.RunConfig,
++       sync_config: ray.tune.RunConfig,
++       failure_config: ray.tune.RunConfig,
++   )
+)
+```
+
+</td>
+<td>
+
+These API imports will be moved in order to decouple the libraries.
+
+</td>
+  </tr>
+
+  <tr>
+<td>🔀</td>
+<td>
+
+```diff
+-ray.train.report(...)
++ray.tune.report(...)
+```
+
+</td>
+<td>
+
+Same as above.
+
+</td>
+  </tr>
+
+  <tr>
+<td>🔀</td>
+<td>
+
+```diff
+-ray.train.get_checkpoint(...)
++ray.tune.get_checkpoint(...)
+```
+
+</td>
+<td>
+
+Same as above.
+
+</td>
+  </tr>
+
+  <tr>
+<td>🔀</td>
+<td>
+
+```diff
+-ray.train.Checkpoint
++ray.tune.Checkpoint
+```
+
+</td>
+<td>
+
+Same as above.
+
+</td>
+  </tr>
+
+  <tr>
+<td>🔀</td>
+<td>
+
+```diff
+-context: TrainContext = ray.train.get_context()
++context: TuneContext = ray.tune.get_context()
+
+-context.get_world_size()
+-context.get_local_world_size()
+-context.get_world_rank()
+-context.get_local_rank()
+-context.get_node_rank()
+context.get_trial_dir()
+context.get_trial_id()
+context.get_trial_name()
+context.get_trial_resources()
+context.get_experiment_name()
+context.get_storage()
+```
+
+</td>
+<td>
+
+Same as above. All of the multi-worker distributed training context information
+do not apply to Tune trials and will be removed.
+
+</td>
+  </tr>
+
+  <tr>
+<td>🔀</td>
+<td>
+
+```diff
+result_grid = tuner.fit()
+-result: ray.train.Result = result_grid[0]
++result: ray.tune.Result = result_grid[0]
+```
+
+</td>
+<td>
+
+Same as above.
+
+</td>
+  </tr>
+
+</table>
+
+## Ray Train Architecture Overview
+
+![Train V2 Architecture](train_architecture.png)
+
+### Controller
+
+The **Controller** is the central orchestrator that manages the distributed execution of training jobs. Its main responsibilities include:
+
+* **Execution Control**:
+  * It triggers the distributed training function across the workers in the `WorkerGroup` and shuts down the workers when training has either finished or failed.
+* **Polling & Health Checking**:
+  * The Controller regularly polls the workers to assess their health and training progress. Based on the worker statuses, it can handle errors and make scaling decisions.
```

Comment:
This looks awesome. A few questions.
Training health monitoring is becoming more and more important and can be quite complex and compute intensive (not just looking at lots of metrics, but also running evals). Where does this logic live?

If the health monitoring takes a while to run but eventually decides that we need to restart from an earlier checkpoint, how does that get implemented?

Reply:
> Where does this logic live?

This logic lives in the Train driver right now, where the controller actor periodically pings the worker actor tasks and handles any errors raised by those health-check tasks.

> If the health monitoring takes a while to run but eventually decides that we need to restart from an earlier checkpoint, how does that get implemented?

Health-checks are every few seconds at the moment, so the decisions happen pretty quickly after Ray detects that a node/actor has died. We also handle more extreme edge-cases where the underlying RPC times out randomly, in which case we'll wait for some configurable time until restarting from the latest checkpoint.

> (not just looking at lots of metrics, but also running evals)
Oh, maybe this part was a bit misleading. At the moment, the "health and training progress" just refers to the underlying actor/node health and reported checkpoints, rather than the health of the training jobs (we do NOT implement anything to handle how well things are converging, how validation metrics are doing, etc).